File: ProductExceptItself.java
//The algorithm initializes an array res and performs a left pass to calculate the product of elements to the left of each index. 
//It then performs a right pass to update res with the product of elements to the right. The final res array contains the product of all elements except the one at each index. 
//The time complexity is O(n), Space - O(1).
class Solution {
    public int[] productExceptSelf(int[] nums) {
        
        int n = nums.length;
        int [] res = new int[n];
        res[0] = 1;
        int rp =1;
        for(int i=1;i<n;i++){
            rp = rp* nums[i-1];
            res[i]= rp;
        }
        rp =1;
        for(int i=n-2;i>0;i--){
            rp *= nums[i+1];
            res[i] = res[i]*rp;
        }
        return res;

    }
}

File: ProductExceptItself.py
# Time Complexity : O(n)
# Space Complexity : O(2n)
# Did this code successfully run on Leetcode : Yes

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        leftProduct = [0]*n
        RightProduct = [0]*n
        Result = [0]*n
        leftProduct[0] = 1
        RightProduct[n-1] = 1
        for i in range(1,n):
            leftProduct[i]=leftProduct[i-1]*nums[i-1]
        for i in range(n-2,-1,-1):
            RightProduct[i] = RightProduct[i+1]*nums[i+1]
        for i in range(n):
            Result[i] = leftProduct[i]*RightProduct[i]
        return Result
        
        